docs/generic-filter-system.md

```diff
@@ -0,0 +1,576 @@
+      },
+    },
+    {
+      type: 'multiSelect',
+      param: 'type',
+      label: 'Type',
+      mode: 'client',
+      options: [
+        { label: 'Build', value: 'build' },
+        { label: 'Test', value: 'test' },
+      ],
+    },
+  ];
+
+  // 2. Automatic filtering and URL state management
+  const { filteredData, isFiltered } = useFilteredData(sortedPipelineRuns, filterConfigs);
+
+  // 3. Simple rendering
+  return (
+    <>
+      <FilterToolbar filterConfigs={filterConfigs} />
+      <Table data={filteredData} /* ... other props */ />
+    </>
+  );
+};
+```
+   No additional components or utilities needed!
+  - No FilterContext wrapper required
+  - No custom filter toolbar component
+  - No manual option generation utilities
+  - No custom filter state types
+
+## Proposed Usage Examples
+
+### API-Mode Filtering
+
+For server-side filtering, the proposed system would use `mode: 'api'`:
+
+```typescript
+const MyAPIListComponent = ({ data }) => {
+  const filterConfigs = [
+    {
+      type: 'search',
+      param: 'name',
+      label: 'Name',
+      mode: 'api',
+    },
+  ];
+
+  // Hook only reads URL parameters, doesn't filter data
+  const { filterValues, isFiltered } = useFilteredData(data, filterConfigs);
+  
+  // Use filterValues to make API calls
+  const filteredData = useK8sWatchResource(Model, { matchLabel: createLabelsFromFilterValue(filterValues) });
+
+  return (
+    <>
+      <FilterToolbar filterConfigs={filterConfigs} />
+      <MyTable data={filteredData} />
+    </>
+  );
+};
+```
+
+### Proposed Custom Filter Functions
+
+```typescript
+const MyClientListComponent = ({ data }) => {
+  const filterConfigs = [
+    {
+      type: 'search',
+      param: 'name',
+      label: 'Name',
+      mode: 'client',
+      filterFn: (item, value) => item.name.toLowerCase().includes(value.toLowerCase()),
+    },
+  ];
+
+  // Proposed hook API
+  const { filteredData, isFiltered } = useFilteredData(data, filterConfigs);
+
+  return (
+    <>
+      <FilterToolbar filterConfigs={filterConfigs} />
+      <MyTable data={filteredData} />
+    </>
+  );
+};
+```
+
+## Filter Configuration
+
+### Options Configuration
+
+You can configure filter options in three ways. Choose **one** approach per filter:
+
+#### 1. Static Options (Known in Advance)
+
+```typescript
+{
+  type: 'multiSelect',
+  param: 'priority',
+  label: 'Priority',
+  mode: 'client',
+  options: [
+    { label: 'High Priority', value: 'high', count: 5 },
+    { label: 'Medium Priority', value: 'medium', count: 12 },
+    { label: 'Low Priority', value: 'low', count: 8 },
+  ],
+}
+```
+
+#### 2. Dynamic Options from Data (Simple Field Extraction)
+
+```typescript
+{
+  type: 'multiSelect',
+  param: 'tags',
+  label: 'Tags',
+  mode: 'client',
+  optionsSource: {
+    from: 'data',
+    key: 'metadata.labels.tag',
+    labelFormatter: (value) => `Tag: ${value}`,
+  },
+}
+```
+
+#### 3. Custom Options Function (Complex Logic) [TODO]: Do we need this?
```

Comment:
maybe we could consider it for phase 2 or 3?

Reply:
Yeah, we can consider for phase 2, I wanted to have this because In some places we  generate filters options from the data.